### PR TITLE
fix: String::split recursion, ends_with/last_index_of shadow-dispatch (#783), chained calls, bool→string (#784)

### PR DESCRIPTION
**Pull request by mizchi**
Four compiler/prelude correctness fixes plus 6 recovered example tests (allowlist 304→333). Each compiler change went through the full battery: bundle regen → `stage3_equal_stage2=true` → full allowlist regression (0 regressions).

## Fixes

**`String::split` infinite recursion** (`0ce675c`)
`prelude/string.vibe`'s `String::split` was a stub that called itself (`String::split(s, sep)`), so any program importing the module and calling split trapped with infinite recursion. Reimplemented via `string_index_of_from` + slicing (empty-separator returns `[s]`). Recovered `examples/parse_test.vibe` (imports the `Int::` prelude helpers).

**#783 — imported `String::ends_with` / `String::last_index_of` trap** (`f6ce194`)
`compile_call.vibe` had unguarded special-case branches emitting **raw-ABI** args and calling `resolve_func(name)`. Correct for the hand-written builtin, but when a prelude function shadows the name it resolves to a value-world (ADR-0055 tagged) ABI function, and raw args trap at runtime. Guarded both with `!prefer_bound_call` (mirroring `String::to_bytes`) so a shadowing definition routes through the normal value-ABI path. Recovered `examples/string_edge_test.vibe` + `examples/string_test.vibe`.

**Parser: chained call suffix after a call/index result** (`8d1a9f9`)
`f()()`, `f()()()`, and `arr[i](x)` failed to parse ("expected ) but got ("). `parse_postfix` stopped when the callee was already an `ECall`. That guard was meant to prevent `f(1)`⏎`(2)` merging across a statement boundary — but the lexer already handles that by inserting a synthetic `;` for newline-then-`(` after an expr-ending token, so the guard only ever blocked legitimate same-line chained calls. Removed that one arm. Recovered `examples/closure_test.vibe` + `examples/closure_advanced_test.vibe`.

**#784 — bool interpolation printed "1"/"0"** (`d8d4f53`)
Interpolation desugars `"\{x}"` → `__to_string(x)` before type-checking, `__to_string` codegen is type-erased, and Bool shares Int's tagged-int runtime rep. Added a syntactic `cc_expr_is_boolish` (mirroring `cc_expr_is_floatish`) that routes `__to_string(<bool expr>)` to a synthesized `if b { "true" } else { "false" }`. Never fires for int args, so the compiler's pervasive int stringification is untouched. Added `examples/bool_to_string_test.vibe`.

## Also confirmed

The RC escaping-closure UAF (previously flagged as a candidate) is **already fixed** — the capture fixtures pass under `VIBE_RC=1` and `VIBE_RC=shadow`.

## Verification

Every compiler change independently: `check_selfhost_bundle_sync` + `check_selfhost_module_source_sync` ok, `stage3_equal_stage2=true`, and full `selfhost_unit_test_runner.sh` regression. Final state: **333/333 allowlisted unit-test files pass, 0 regressions**. The statement-boundary case `f(1)`⏎`(2)` and int/string/float interpolation were explicitly re-verified unchanged.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Wo3ZjDzboj8Ps41rv2Qa9K

---
_Generated by [Claude Code](https://claude.ai/code/session_01Wo3ZjDzboj8Ps41rv2Qa9K)_

### CODE REVIEW

File: lib/@vibe/compiler/codegen/expr/compile_call.vibe [hunk 1]
```diff
@@ -1441,6 +1471,12 @@ export let compile_call: (Bytes, Expr, Array[Expr], Array[String], Int, CompileC
       // bits (or the box pointer under RC) as an integer.
       ce(buf, double_to_string_expr(Array::get(args, 0)), local_names, next_local, ctx, ld)
     }
+    else if fname == "__to_string" && Array::length(args) == 1 && cc_expr_is_boolish(Array::get(args, 0)) {
+      // #784: a statically-boolean argument formats as "true"/"false", not the
+      // integer 1/0 the generic path below would emit. Reuse the existing
+      // if/EString codegen: `if b { "true" } else { "false" }`.
+      ce(buf, EIf(Array::get(args, 0), EString("true"), EString("false")), local_names, next_local, ctx, ld)
```
[the remark this answers is not in the document]
Comment (mizchi) [reply lib/@vibe/compiler/codegen/expr/compile_call.vibe:1477]: Confirmed and fixed in 422870e. Reproduced: `"\{1 < 2}"` in a file with no `"true"`/`"false"` literal failed with `string not in table: true`. `collect_strings_expr` now seeds both literals whenever `__to_string` is called. Added `examples/bool_interp_no_literal_test.vibe` (asserts via `String::length` so the file contains neither literal, exercising the seeding). Full regression 334/334, stage3==stage2.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lib/@vibe/compiler/codegen/expr/compile_call.vibe [hunk 2]
```diff
@@ -171,6 +171,36 @@ let rec cc_expr_is_floatish: (Expr, Array[String], CompileCtx) -> Bool with { Er
...
+  } else if tag == 5 {
+    let op = get_ebinop_op(e)
+    op == "==" || op == "!=" || op == "<" || op == ">" || op == "<=" || op == ">=" || op == "&&" || op == "||"
+  } else if tag == 6 {
+    get_eunaryop_op(e) == "!"
+  } else if tag == 8 {
+    let callee = get_ecall_callee(e)
+    if is_eident(callee) {
+      let fname = get_eident_name(callee)
+      fname == "eq" || fname == "not" || fname == "String::equals" || fname == "String::contains" || fname == "String::starts_with" || fname == "String::ends_with" || fname == "Array::any" || fname == "Array::all" || fname == "Array::contains" || fname == "Map::has_key" || fname == "map_has" || fname == "Path::is_absolute" || fname == "Fs::exists" || fname == "Fs::is_dir" || fname == "Fs::is_file"
```
[the remark this answers is not in the document]
Comment (mizchi) [reply lib/@vibe/compiler/codegen/expr/compile_call.vibe]: Good catch, fixed in 422870e. There's no cheap sound signal to make the name match shadow-aware (those builtins live in the func_table too, so `prefer_bound_call` is true either way), so I dropped the bool-builtin-call case entirely. `cc_expr_is_boolish` now classifies only UNSHADOWABLE forms — bool literal, comparison/logical binary op, `!` unary (operators bypass name resolution). Verified: `let not = (x: Int) -> Int { x + 100 }` then `"\{not(5)}"` now prints "105", not "true". `"\{String::contains(...)}"` reverts to the 1/0 limitation shared with variable-bound bools.

---
_Generated by [Claude Code](https://claude.ai/code)_
